From c67295d95d1c581e9490587fb154a45e7b606d4a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Wed, 2 Apr 2025 00:51:06 +0000
Subject: [PATCH] window: Fix error handling in case of decoding failure

When chardet cannot detect the encoding, it returns None, which needs to
be handled as error. Also, give a more detailed error message to the
user, and reset the UI by calling new_document().
---
 apostrophe/main_window.py | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/apostrophe/main_window.py b/apostrophe/main_window.py
index c365c7e..64ca8ac 100644
--- a/apostrophe/main_window.py
+++ b/apostrophe/main_window.py
@@ -553,18 +553,27 @@ class MainWindow(Adw.ApplicationWindow):
         except GLib.GError as error:
             helpers.show_error(self, str(error.message))
             LOGGER.warning(str(error.message))
+            self.new_document()
             return
 
         try:
             try:
                 self.current.encoding = 'UTF-8'
                 decoded = contents.decode(self.current.encoding)
-            except UnicodeDecodeError:
+            except UnicodeDecodeError as error:
                 self.current.encoding = chardet.detect(contents)['encoding']
+                if not self.current.encoding:
+                    raise error
                 decoded = contents.decode(self.current.encoding)
         except UnicodeDecodeError as error:
-            helpers.show_error(self, str(error.message))
-            LOGGER.warning(str(error.message))
+            helpers.show_error(
+                self,
+                _("Failed to detect encoding for file:\n"
+                "{file}\n\nThe error was:\n{err_msg}")
+                .format(file=gfile.get_path(),
+                        err_msg=str(error).encode().decode("unicode-escape")))
+            LOGGER.warning(str(error))
+            self.new_document()
             return
         else:
             self.textview.set_text(decoded)
-- 
GitLab

